### ISSUE TITLE
[ISSUE/BUG] Error: not enough slots available in the system

### ISSUE THREAD

**Issue by gopikrishnangs44** [geoschem/gchp#254]
I have installed GCHP 13.4.0 through spack through the following commands

spack install gcc@9.3.0
spack compiler find $(spack location -i gcc@9.3.0)
spack install gchp%gcc@9.3.0 ^openmpi +pmi schedulers=slurm
export MPI_ROOT=$(spack location -i openmpi)
spack load gchp

Created a run directory using ./createRunDir.sh

Edited runConfig.sh

TOTAL_CORES=6
NUM_NODES=1
NUM_CORES_PER_NODE=6

Ran ./runConfig.sh

Copied the executable to the run directory and ran ./gchp,

Produced the error:

 adding HEMCO export: EmisSOAP_Total
 adding HEMCO export: EmisSOAP_Aircraft
 adding HEMCO export: EmisSOAP_Anthro
 adding HEMCO export: EmisSOAP_Biogenic
 adding HEMCO export: EmisSOAP_BioBurn
 adding HEMCO export: EmisSOAP_Ship
 adding HEMCO export: EmisSOAS_Biogenic
 adding HEMCO export: EmisTOLU_Total
 adding HEMCO export: EmisTOLU_Anthro
 adding HEMCO export: EmisTOLU_BioBurn
 adding HEMCO export: EmisTOLU_Ship
 adding HEMCO export: EmisXYLE_Total
 adding HEMCO export: EmisXYLE_Anthro
 adding HEMCO export: EmisXYLE_BioBurn
 adding HEMCO export: EmisXYLE_Ship
 adding HEMCO export: HcoLightningFlashRate_Total
 adding HEMCO export: HcoLightningFlashRate_IntraCld
 adding HEMCO export: HcoLightningFlashRate_CldGround
 adding HEMCO export: HcoConvectiveCloudTopHeight
        GENERIC: INFO: Finished the 'SetServices' stage of the gridded component 'GCHPchem'
        GENERIC: INFO: Started the  'SetServices' stage of the gridded component 'DYNAMICS'
 Character Resource Parameter: DYCORE:OFF
NOTE: MPP_DOMAINS_SET_STACK_SIZE: stack size set to    32768.
&MPP_IO_NML
 HEADER_BUFFER_VAL=16384      ,
 GLOBAL_FIELD_ON_ROOT_PE=T,
 IO_CLOCKS_ON=F,
 SHUFFLE=0          ,
 DEFLATE_LEVEL=-1         ,
 CF_COMPLIANCE=F,
 /
NOTE: MPP_IO_SET_STACK_SIZE: stack size set to     131072.
NOTE: MPP_DOMAINS_SET_STACK_SIZE: stack size set to 20000000.
 Integer*4 Resource Parameter: NX:1
 Integer*4 Resource Parameter: NY:6
 Integer*4 Resource Parameter: IM:24
 Integer*4 Resource Parameter: JM:144
 Integer*4 Resource Parameter: RUN_DT:600
 For k_split (remapping)=           1
n_split is set to 02 for resolution-dt=0025x0025x6-  600.000
Using n_zfilter : 000
Using n_sponge : 001
Using non_ortho :       T
 Starting PEs :            1
 Starting Threads :            1

FATAL: mpp_domains_define.inc: not all the pe_end are in the pelist

FATAL: mpp_domains_define.inc: not all the pe_end are in the pelist

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.

**Comment by lizziel** [comment]
Hi @gopikrishnangs44, you need to run GCHP using at least 6 cores. Could you detail why you ran GCHP using the command `./gchp`? This is not the proper way to submit GCHP.

**Comment by gopikrishnangs44** [comment]
Deare @lizziel, 
It was an openmpi issue. I was not aware of the srun option. The issue is solved and I am able to run GCHP. Please try to detail the use of srun in documentation. Also, I would suggest to add a comment regarding the usage of srun option on the video tutorial by @LiamBindle. 
Thank you for your time.

**Comment by gopikrishnangs44** [comment]
I had some issues in srun.

Therefore I installed mpirun through spack using 'openmpi+legacylaunchers'

Changed runConfig.sh
#runConfig.sh
TOTAL_CORES=36
NUM_NODES=1
NUM_CORES_PER_NODE=36

**GCHP is running perfectly with 1 node and 36 processors using the batch-script:**

#!/bin/bash
#SBATCH -J gc # name of the job
#SBATCH -p standard-low # name of the partition: available options "standard,standard-low,gpu,gpu-low,hm"
#SBATCH -n 36 #no of processes or tasks
#SBATCH -N 1 # Number of nodes
#SBATCH --cpus-per-task=1 # no of threads per process or task
#SBATCH -t 24:00:00 # walltime in HH:MM:SS, Max value 72:00:00

source /home/21cl91p01/spack/share/spack/setup-env.sh
spack load gchp
spack load /gmukwis                                       #openmpi with legacy launchers.
sh /scratch/21cl91p01/test-run/runConfig.sh
mpirun -np 36 gchp
                    

**But when I increased the number of Nodes (N); I am facing an error. for instance:** 

#runConfig.sh
TOTAL_CORES=36
NUM_NODES=2
NUM_CORES_PER_NODE=18

 and 

#!/bin/bash
#SBATCH -J gc # name of the job
#SBATCH -p standard-low # name of the partition: available options "standard,standard-low,gpu,gpu-low,hm"
#SBATCH -n 36 #no of processes or tasks
#SBATCH -N 2 # Number of nodes
#SBATCH --cpus-per-task=1 # no of threads per process or task
#SBATCH -t 24:00:00 # walltime in HH:MM:SS, Max value 72:00:00

source /home/21cl91p01/spack/share/spack/setup-env.sh
spack load gchp
spack load /gmukwis                                       #openmpi with legacy launchers.
sh /scratch/21cl91p01/test-run/runConfig.sh
mpirun -np 36 gchp

**I am getting the following error:**


SLURM_CLUSTER_NAME = param-shakti
SLURM_JOB_ACCOUNT = jayanarayanancl
SLURM_JOB_ID = 936426
SLURM_JOB_NAME = gc
SLURM_JOB_NODELIST = cn[108-109]
SLURM_JOB_USER = 21cl91p01
SLURM_JOB_UID = 6687
SLURM_JOB_PARTITION = standard-low
SLURM_TASK_PID = 143901
SLURM_SUBMIT_DIR = /scratch/21cl91p01/test-run
SLURM_CPUS_ON_NODE = 18
SLURM_NTASKS = 36
SLURM_TASK_PID = 143901


There are not enough slots available in the system to satisfy the 36
slots that were requested by the application:

  gchp

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the Open MPI term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which Open MPI processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, Open MPI defaults to the number of processor cores

In all the above cases, if you want Open MPI to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --oversubscribe option to ignore the
number of available slots when deciding the number of processes to
launch.

**The same error repeats in different cases, when I tried 144 process with 4 nodes, 72 process with 2 nodes etc.**

**Comment by lizziel** [comment]
@Jourdan-He and @SaptSinha, could you test out istalling GCHP and its dependencies with spack to see if you can reproduce this issue?

**Comment by gopikrishnangs44** [comment]
I have the same issue while running GCHP 13.0.1.

**Comment by lizziel** [comment]
Hi @gopikrishnangs44, @Jourdan-He on the support team will try building and running GCHP with Spack to see if the issue is reproducible.

**Comment by Jourdan-He** [comment]
Hi, I have tried intalling and running GCHP 13.4.1 through a Spack container and it worked fine. The container I used is: jourdanhe/gchpspack:latest.

**Comment by lizziel** [comment]
@gopikrishnangs44, since this is not a problem with the spack libraries I suggest you contact your compute cluster administrator to help figure out what the problem is with how you are submitting your job. I found a few hits on the internet when I searched your issue, e.g. [this](https://stackoverflow.com/questions/35704637/mpirun-not-enough-slots-available), and you could try altering how to submit jobs to your cluster based on suggestions there.

**Comment by gopikrishnangs44** [comment]
@lizziel Can you please edit the link for example you gave, its not working. I will try that

**Comment by gopikrishnangs44** [comment]
mpirun --oversubscribe -np 36 gchp 
command solved the issue.

**Comment by lizziel** [comment]
Hi @gopikrishnangs44, I'm glad you were able to find the correct link and get it working using the advice there. I fixed the link above so if anyone else runs into this issue they can do this as well. I'll also update the title. Thanks for following up!

**Comment by gopikrishnangs44** [comment]
Please note that the Accepted answer in the above [link](https://stackoverflow.com/questions/35704637/mpirun-not-enough-slots-available) didn't worked in our HPC Cluster. 

However,  --oversubscribe worked.

Thank you @lizziel for your timely responses indeed.

**Comment by Twize** [comment]
Sorry for commenting on a closed thread, but @gopikrishnangs44 just curious, did the issue you have with srun have something to do with Munge? I installed GCHP's dependencies using Spack, and tried running with srun, however I got the following error: 

srun: error: If munged is up, restart with --num-threads=10
srun: error: Munge encode failed: Failed to access "/gpfs/fs1/home/d/dylan/tylerw/mymodules/spack/opt/sp$
srun: error: slurm_send_node_msg: auth_g_create: REQUEST_HET_JOB_ALLOC_INFO has authentication error
srun: error: Unable to confirm allocation for job 8504167: Protocol authentication error
srun: Check SLURM_JOB_ID environment variable. Expired or invalid job 8504167

I am trying the same approach of installing mpirun at the moment using spack and 'openmpi+legacylaunchers' to see if it runs with that instead.
